File: C-Abi-Bright-Aot/Tests/CAbiBridge.Aot.Vb.Tests/TestCrypto/CryptoTestUtils.vb
Option Strict On
Option Explicit On

Imports System.IO
Imports michele.natale.CAbiBridge
Imports System.Runtime.InteropServices
Imports System.Text
Imports System.Numerics
Imports System.Runtime.CompilerServices

Namespace michele.natale.Tests
  Friend Module CryptoTestUtils

    Friend Sub SetRngFileData(filename As String, size As Int32)
      Using fsout = New FileStream(filename, FileMode.Create, FileAccess.Write)
        Dim length = If(size < 1024 * 1024, size, 1024 * 1024)

        While length > 0
          fsout.Write(RngBytes(length))
          size -= length
          length = If(size < 1024 * 1024, size, 1024 * 1024)
        End While
      End Using
    End Sub

    Friend Async Function SetRngFileDataAsync(filename As String, size As Int32) As Task
      Dim bufferSize = 1024 * 1024

      Dim options = New FileStreamOptions With
      {
       .Mode = FileMode.Create, .Access = FileAccess.Write,
       .Share = FileShare.None, .BufferSize = bufferSize,
       .Options = FileOptions.Asynchronous
      }

      Using fsout As New FileStream(filename, options)

        While size > 0
          Dim chunk = Math.Min(size, bufferSize)
          Dim data = RngBytes(chunk)

          Await fsout.WriteAsync(data.AsMemory(0, chunk))

          size -= chunk
        End While

      End Using
    End Function


    Friend Function RngBytes(size As Int32) As Byte()
      Dim rand = Random.Shared
      Dim result = New Byte(size - 1) {}
      rand.NextBytes(result)
      If result(0) = 0 Then result(0) += CByte(1)
      Return result
    End Function

    Friend Sub AssertError(err As CError)
      If err.error_code <> 0 Then
        Dim msg = Marshal.PtrToStringAnsi(err.message)
        Throw New ArgumentException(msg)
      End If
    End Sub

    Friend Function ToBytes(bytes As IntPtr, length As Int32) As Byte()
      Dim result = New Byte(length - 1) {}
      Marshal.Copy(bytes, result, 0, length)
      Return result
    End Function


    Friend Function ToInts(Of T As INumber(Of T))(ptr As IntPtr, length As Int32) As T()
      'For uint8 to uint64
      Dim szt = Unsafe.SizeOf(Of T)()
      Dim bytes = ToBytes(ptr, length * szt)

      Dim result = New T(length - 1) {}
      Buffer.BlockCopy(bytes, 0, result, 0, length * szt)

      Return result
    End Function

    Public Function ToBools(ptr As IntPtr, length As Int32) As Boolean()
      Dim result = New Boolean(length - 1) {}
      For i = 0 To length - 1
        result(i) = Marshal.ReadByte(ptr, i) <> 0
      Next

      Return result
    End Function

    Friend Function ToFloats(Of T As IFloatingPoint(Of T))(ptr As IntPtr, length As Int32) As T()
      'For double and float
      Dim tsz As Int32 = Unsafe.SizeOf(Of T)() * length

      Dim bytes = New Byte(tsz - 1) {}
      Marshal.Copy(ptr, bytes, 0, tsz)

      Dim result = New T(length - 1) {}
      Buffer.BlockCopy(bytes, 0, result, 0, tsz)

      Return result
    End Function

    Friend Function ToDecimals(ptr As IntPtr, length As Int32) As Decimal()
      Dim result = New Decimal(length - 1) {}

      For i = 0 To length - 1
        Dim offset = i * 16

        Dim lo = Marshal.ReadInt32(ptr, offset + 0)
        Dim mid = Marshal.ReadInt32(ptr, offset + 4)
        Dim hi = Marshal.ReadInt32(ptr, offset + 8)
        Dim flags = Marshal.ReadInt32(ptr, offset + 12)

        result(i) = New Decimal(lo, mid, hi, (flags And &H80000000) <> 0, CByte(flags >> 16 And &H7F))
      Next

      Return result
    End Function


    'Friend Function ToCharsUtf8(ptr As IntPtr, length As Integer) As Char()
    '  Dim bytes = New Byte(length - 1) {}
    '  Marshal.Copy(ptr, bytes, 0, length)

    '  Return Encoding.UTF8.GetChars(bytes)
    'End Function

    'Friend Function ToCharsAscii(ptr As IntPtr, length As Integer) As Char()
    '  Dim result = New Char(length - 1) {}
    '  For i = 0 To length - 1
    '    result(i) = CChar(Marshal.ReadByte(ptr, i))
    '  Next

    '  Return result
    'End Function

    'Friend Function ToCharsUtf16(ptr As IntPtr, length As Integer) As Char()
    '  Dim result = New Char(length - 1) {}
    '  For i = 0 To length - 1
    '    result(i) = CChar(Marshal.ReadInt16(ptr, i * 2))
    '  Next

    '  Return result
    'End Function

  End Module
End Namespace

File: C-Abi-Bridge-Aot/Tests/CAbiBridgeAot.Tests/TestCrypto/CryptoTestUtils.cs



using System.Text;
using System.Numerics;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace michele.natale.Tests;

using CAbiBridge;

internal class CryptoTestUtils
{
  internal static void SetRngFileData(string filename, int size)
  {
    using var fsout = new FileStream(
      filename, FileMode.Create, FileAccess.Write);

    var length = size < 1024 * 1024 ? size : 1024 * 1024;

    while (length > 0)
    {
      fsout.Write(RngBytes(length));
      size -= length; length = size < 1024 * 1024 ? size : 1024 * 1024;
    }
  }

  internal async static Task SetRngFileDataAsync(
    string filename, int size)
  {
    await using var fsout = new FileStream(
      filename, FileMode.Create, FileAccess.Write);

    var length = size < 1024 * 1024 ? size : 1024 * 1024;

    while (length > 0)
    {
      fsout.Write(RngBytes(length));
      size -= length; length = size < 1024 * 1024 ? size : 1024 * 1024;
    }
  }

  internal static byte[] RngBytes(int size)
  {
    //in practice, using a crypto-random 
    var rand = Random.Shared;
    var result = new byte[size];
    rand.NextBytes(result);
    if (result[0] == 0) result[0]++;
    return result;
  }

  internal static void AssertError(CError err)
  {
    if (err.error_code != 0)
    {
      var msg = Marshal.PtrToStringAnsi(err.message);
      throw new ArgumentException(msg);
    }
  }

  internal static byte[] ToBytes(
    IntPtr bytes, int length)
  {
    var result = new byte[length];
    Marshal.Copy(bytes, result, 0, length);
    return result;
  }

  internal static T[] ToInts<T>(
    IntPtr ptr, int length)
    where T : INumber<T>
  {
    //For uint8 to uint64
    var szt = Unsafe.SizeOf<T>();
    var bytes = ToBytes(ptr, length * szt);

    var result = new T[length];
    Buffer.BlockCopy(bytes, 0, result, 0, length * szt);

    return result;
  }

  public static bool[] ToBools(
    IntPtr ptr, int length)
  {
    var result = new bool[length];
    for (int i = 0; i < length; i++)
      result[i] = Marshal.ReadByte(ptr, i) != 0;

    return result;
  }

  internal static T[] ToFloats<T>(
    IntPtr ptr, int length)
    where T : IFloatingPoint<T>
  {
    //For double and float
    int tsz = Unsafe.SizeOf<T>() * length;

    var bytes = new byte[tsz];
    Marshal.Copy(ptr, bytes, 0, tsz);

    var result = new T[length];
    Buffer.BlockCopy(bytes, 0, result, 0, tsz);

    return result;
  }

  internal static decimal[] ToDecimals(IntPtr ptr, int length)
  {
    var result = new decimal[length];

    for (int i = 0; i < length; i++)
    {
      int offset = i * 16;

      int lo = Marshal.ReadInt32(ptr, offset + 0);
      int mid = Marshal.ReadInt32(ptr, offset + 4);
      int hi = Marshal.ReadInt32(ptr, offset + 8);
      int flags = Marshal.ReadInt32(ptr, offset + 12);

      result[i] = new decimal(
        lo, mid, hi, (flags & unchecked((int)0x80000000)) != 0,
          (byte)((flags >> 16) & 0x7F));
    }

    return result;
  }


  internal static char[] ToCharsUtf8(IntPtr ptr, int length)
  {
    var bytes = new byte[length];
    Marshal.Copy(ptr, bytes, 0, length);

    return Encoding.UTF8.GetChars(bytes);
  }

  internal static char[] ToCharsAscii(IntPtr ptr, int length)
  {
    var result = new char[length];
    for (int i = 0; i < length; i++)
      result[i] = (char)Marshal.ReadByte(ptr, i);

    return result;
  }

  internal static char[] ToCharsUtf16(IntPtr ptr, int length)
  {
    var result = new char[length];
    for (int i = 0; i < length; i++)
      result[i] = (char)Marshal.ReadInt16(ptr, i * 2);

    return result;
  }
}
